File: Python/Algorithms/917.py
class Solution:
    alpha = {*ascii_letters}
    
    def reverseOnlyLetters(self, S):
        alpha = self.alpha
        start, end = 0, len(S) - 1
        S = list(S)
        while start < end:
            while S[start] not in alpha:
                if start == end:
                    return "".join(S)
                start = start + 1
            while S[end] not in alpha:
                if start == end:
                    return "".join(S)
                end = end - 1
            S[start], S[end] = S[end], S[start]
            start, end = start + 1, end - 1
        return "".join(S)

File: Rust/Algorithms/917.rs
use std::collections::HashSet;

impl Solution {
    pub fn reverse_only_letters(s: String) -> String {
        if s.len() == 0 {
            return s;
        }
        let letters: HashSet<_> =
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
                .chars()
                .collect();
        let (mut start, mut end) = (0, s.len() - 1);
        let mut S: Vec<_> = s.chars().collect();

        while start < end {
            while !letters.contains(&S[start]) {
                if start == end {
                    return S.iter().collect::<String>();
                }
                start += 1;
            }
            while !letters.contains(&S[end]) {
                if start == end {
                    return S.iter().collect::<String>();
                }
                end -= 1;
            }
            let tmp = S[start];
            S[start] = S[end];
            S[end] = tmp;
            start += 1;
            end -= 1;
        }
        S.iter().collect::<String>()
    }
}
